src/filters/method.rs

```diff
@@ -137,3 +137,144 @@ where
+    //! the request `Method`, and if not matched, will reject the request with a
+    //! `405 Method Not Allowed`.
+    //!
+    //! These "filters" behave a little differently than the rest. Instead of
+    //! being used directly on requests, these filters "wrap" other filters.
+    //!
+    //!
+    //! ## Wrapping a `Filter` (`with`)
+    //!
+    //! ```
+    //! use warp::Filter;
+    //! use warp::filters::method::v2 as methodv2;
+    //!
+    //! let route = warp::any()
+    //!     .map(warp::reply);
+    //!
+    //! // This filter rejects any non-GET requests before they get to the
+    //! // wrapped `route`.
+    //! let get_only = route
+    //!     .with(methodv2::get());
+    //! ```
+    //!
+    //! Wrapping allows adding in conditional logic *before* the request enters
+    //! the inner filter. In the example, it rejects non-GET requests before
+    //! they get to the wrapped `route` filter.
+    use futures::{future::FutureResult, IntoFuture};
+    use http::Method;
+
+    use filter::{And, Filter, FilterBase, MapErr, WrapSealed};
+    use reject::{CombineRejection, Rejection};
+    use reply::Reply;
+
+    /// Wrap a `Filter` to require that the request method to be `GET`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use warp::Filter;
+    /// use warp::filters::method::v2 as methodv2;
+    ///
+    /// let route = warp::any().map(warp::reply);
+    /// let get_only = route.with(methodv2::get());
+    /// ```
+    pub fn get() -> WithMethod {
+        WithMethod(&Method::GET)
+    }
+
+    /// Wrap a `Filter` to require that the request method to be `POST`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use warp::Filter;
+    /// use warp::filters::method::v2 as methodv2;
+    ///
+    /// let route = warp::any().map(warp::reply);
+    /// let post_only = route.with(methodv2::post());
+    /// ```
+    pub fn post() -> WithMethod {
+        WithMethod(&Method::POST)
+    }
+
+    /// Wrap a `Filter` to require that the request method to be `PUT`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use warp::Filter;
+    /// use warp::filters::method::v2 as methodv2;
+    ///
+    /// let route = warp::any().map(warp::reply);
+    /// let put_only = route.with(methodv2::put());
+    /// ```
+    pub fn put() -> WithMethod {
+        WithMethod(&Method::PUT)
+    }
+
+    /// Wrap a `Filter` to require that the request method to be `DELETE`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use warp::Filter;
+    /// use warp::filters::method::v2 as methodv2;
+    ///
+    /// let route = warp::any().map(warp::reply);
+    /// let delete_only = route.with(methodv2::delete());
+    /// ```
+    pub fn delete() -> WithMethod {
+        WithMethod(&Method::DELETE)
+    }
+
+    /// Wrap a `Filter` to require a specific HTTP method.
+    #[derive(Debug, Clone, Copy)]
+    pub struct WithMethod(&'static Method);
```

Comment:
I don't know how important this is, but I thought I'd explain just for context: the previous `MethodIs` thingy took a closure so that the size of it could be 0, instead of holding a static reference. In this case, `size_of::<WithMethod>()` is 1 word, and cloning (copying) it probably means a copying that pointer? I didn't put any measuring into whether this matters at all, just figured I'd explain the *why* of the older one.

Reply:
Ah, that's useful context. This wasn't even remotely something I was thinking about. If there is a performance impact, I guess these can be implemented as individual filters, one per method, or by using a phantom type + associated const instead of storing the method in the filter struct.